### PR TITLE
HIVE-28262:Single column use MultiDelimitSerDe parse column error

### PR DESCRIPTION
**Pull request by LaughingVzr**
### What changes were proposed in this pull request?
modify LazyStruct#findIndexes function and LazyStruct#parseMultiDelimit function, change fields.length Conditional judgment：
```java
public void parseMultiDelimit(byte[] rawRow, byte[] fieldDelimit) {
     - if (fields.length > 1 && delimitIndexes[i - 1] != -1) {
     + if (delimitIndexes[i - 1] != -1) {
  }

 private int[] findIndexes(byte[] array, byte[] target) {
    - if (fields.length <= 1) {
    + if (fields.length < 1) {
   ...
    - for (int i = 1; i < indexes.length; i++) {
    + for (int i = 1; i <= indexes.length; i++) {
    ...
    }
    return indexes;
  }
```

I add an test for this fix:
```java
@Test
    public void testParseMultiDelimit() throws Throwable {
        try {
            // single field named id
            List<String> structFieldNames = new ArrayList<>();
            structFieldNames.add("id");
            // field type is string
            List<TypeInfo> fieldTypes = new ArrayList<>();
            PrimitiveTypeInfo primitiveTypeInfo = new PrimitiveTypeInfo();
            primitiveTypeInfo.setTypeName("string");
            fieldTypes.add(primitiveTypeInfo);

            // separators + escapeChar => "|"
            byte[] separators = new byte[]{124, 2, 3, 4, 5, 6, 7, 8};

            // sequence =>"\N"
            Text sequence = new Text();
            sequence.set(new byte[]{92, 78});

            // create lazy object inspector parameters
            LazyObjectInspectorParameters lazyObjectInspectorParameters = new LazyObjectInspectorParametersImpl(false, (byte) '0',
                    false, null, separators, sequence);
            // create a lazy struct inspector
            ObjectInspector lazyStructInspector = LazyFactory.createLazyStructInspector(structFieldNames, fieldTypes, lazyObjectInspectorParameters);
            LazyStruct lazyStruct = (LazyStruct) LazyFactory.createLazyObject(lazyStructInspector);

            // origin row data
            String rowData = "1|@|";
            // row field delimiter
            String fieldDelimiter = "|@|";

            // parse row use multi delimit
            lazyStruct.parseMultiDelimit(rowData.getBytes(StandardCharsets.UTF_8),
                    fieldDelimiter.getBytes(StandardCharsets.UTF_8));

            // check the first field and second field start position index
            // before fix result: 0,1
            // after fix result: 0,2
            Assert.assertArrayEquals(new int[]{0, 2}, lazyStruct.startPosition);
        } catch (Throwable e) {
            e.printStackTrace();
            throw e;
        }
    }
```

### Why are the changes needed?
If a table only have one column field with multidelimit，query this column data is error data.
When I use this data to do other operation(e.g cast use UDFToLong function),get result is NULL.

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
test class: serde/src/test/org/apache/hadoop/hive/serde2/lazy/TestLazyStruct.java
test function: org.apache.hadoop.hive.serde2.lazy.TestLazyStruct#testParseMultiDelimit
test command: mvn test -Dtest=TestLazyStruct --pl serde

### CODE REVIEW

**Review by zhangbutao — COMMENTED** [review]
Hi, @LaughingVzr Thanks for you contribution. Failed tests seemed to be related your change, Please take a look: http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5252/7/tests

**Comment by LaughingVzr** [comment]
> Hi, @LaughingVzr Thanks for you contribution. Failed tests seemed to be related your change, Please take a look: http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5252/7/tests

Ok, I will check it

File: serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
  * the parent LazyObject.
  */
 public class LazyStruct extends LazyNonPrimitive<LazySimpleStructObjectInspector>
-    implements StructObject, SerDeStatsStruct {
+        implements StructObject, SerDeStatsStruct {
```
Comment (zhangbutao) [inline serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java:43]: Remove the unrelated change.

File: serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ private void parse() {
         fieldByteEnd++;
       } else {
         if (isEscaped && bytes[fieldByteEnd] == escapeChar
-            && fieldByteEnd + 1 < structByteEnd) {
+                && fieldByteEnd + 1 < structByteEnd) {
```
Comment (zhangbutao) [inline serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java:141]: Remove the unrelated change.

File: serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java [hunk 3]
```diff
@@ -151,14 +151,14 @@ private void parse() {
     if (!extraFieldWarned && fieldByteEnd < structByteEnd) {
       extraFieldWarned = true;
       LOG.warn("Extra bytes detected at the end of the row! Ignoring similar "
-          + "problems.");
+              + "problems.");
```
Comment (zhangbutao) [inline serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java:154]: Remove the unrelated change.

File: serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java [hunk 4]
```diff
@@ -151,14 +151,14 @@ private void parse() {
...
     }
 
     // Missing fields?
     if (!missingFieldWarned && fieldId < fields.length) {
       missingFieldWarned = true;
       LOG.info("Missing fields! Expected " + fields.length + " fields but "
-          + "only got " + fieldId + "! Ignoring similar problems.");
+              + "only got " + fieldId + "! Ignoring similar problems.");
```
Comment (zhangbutao) [inline serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazyStruct.java:161]: Remove the unrelated change.

**Comment by zhangbutao** [comment]
LGTM.  Pending some nit comments. :)
Thanks for the fix!

**Review by LaughingVzr — COMMENTED** [review]
LGTM. Thanks for your review.

**Review by LaughingVzr — COMMENTED** [review]
LGTM. Thanks for your review.

**Review by zhangbutao — APPROVED** [review]
+1
